notebooks/visit_data.py

```diff
@@ -0,0 +1,255 @@
+# done mask, removes 1.0%
+done_mask = visit_data_reduced['pagePath_parent'] == 'done'
+visit_data_reduced = visit_data_reduced[~done_mask]
+
+fields_of_interest = ['fullVisitorId', 'visitId', 'pagePath', 'pagePath_abbreviated']
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# lag1 page mask, removes 12%
+visit_data_reduced['pagePath_abbreviated_lag1'] = visit_data_reduced['pagePath_abbreviated'].shift()
+page_mask = visit_data_reduced['pagePath_abbreviated'] != visit_data_reduced['pagePath_abbreviated_lag1']
+visit_data_reduced = visit_data_reduced[page_mask]
+visit_data_reduced.shape
+# create a lag of visitor id to ensure that the lag1 page path relates to the same visitor
+# this cuts off the first stage of all journeys (as they have no valid lag 1 page path)
+# reduces the dataset down by a further 28%
+visit_data_reduced['fullVisitorId_lag1'] = visit_data_reduced['fullVisitorId'].shift()
+id_mask = visit_data_reduced['fullVisitorId'] == visit_data_reduced['fullVisitorId_lag1']
+visit_data_reduced = visit_data_reduced[id_mask]
+visit_data_reduced.shape
+visit_data_reduced[['pagePath_abbreviated_lag1', 'pagePath_abbreviated']]
+pd.DataFrame(visit_data_reduced['pagePath_abbreviated_lag1'].value_counts()).head(10)
+pd.DataFrame(visit_data_reduced['pagePath_abbreviated_lag1'].value_counts()).tail(10)
+
+# 34k levels of unique parent pagePath: might make it difficult to use knn, as its sparse
+len(visit_data_reduced['pagePath_abbreviated'].unique())
+
+# do a test train split and predict using the max count cat for the prior page to predict the subsequent page
+unique_visitorId = visit_data_reduced['fullVisitorId'].unique()
+random_df = pd.DataFrame({'fullVisitorId': unique_visitorId,
+                          'random_id': np.random.random(len(unique_visitorId))})
+visit_data_reduced = visit_data_reduced.merge(random_df, on='fullVisitorId')
+threshold_test_train = 0.7
+# train data
+fields_keep = ['date', 'fullVisitorId', 'visitId', 'hour', 'minute',
+               'pagePath_abbreviated', 'pagePath_abbreviated_lag1']
+train_data = visit_data_reduced[visit_data_reduced['random_id'] < threshold_test_train][fields_keep]
+train_data[train_data['fullVisitorId'] == 3250250007886851499]
+
+# test data
+test_data = visit_data_reduced[visit_data_reduced['random_id'] >= threshold_test_train][fields_keep]
+test_data.head()
+
+# build a dictionary of dictionaries for predictions
+'''
+predictions are based on looking up the (pagePath_parent, pagePath_parent_lag1) tuple in the dictionary
+built from the train data, and returning the reciprocal rank of the pagePath_parent
+'''
+
+pagePath_predictor = {}
+pagePath_count = {}
+for p_lag1, p in zip(train_data['pagePath_abbreviated_lag1'], train_data['pagePath_abbreviated']):
+    pagePath_predictor[(p_lag1, p)] = pagePath_predictor.get((p_lag1, p), 0) + 1
+    pagePath_count[p_lag1] = pagePath_count.get(p_lag1, 0) + 1
```

Comment:
Took me awhile to get what this is doing! To check my understanding, this is counting the number of times `(p_lag1, p)` are in `zip(train_data['pagePath_abbreviated_lag1'], train_data['pagePath_abbreviated'])`, right?